### ISSUE TITLE
Yarn installs nested dependency when should not

### ISSUE THREAD

**Issue by wclr** [yarnpkg/yarn#6842]
Hm, just don't expect such strange not expected behavior of yarn. This seems really strange and severe.

**What is the current behavior?**

Say have such deps:
```
"dependencies": {
  "webapp-webpack-plugin": "2.3.0",
  "favicons": "5.1.1"
}
```

`webapp-webpack-plugin@2.3.0` package has dependency `"favicons": "^5.1.1",`

Yarn installs the latest version of `favicons@5.3.0` in the `webapp-webpack-plugin/node_modules/favicons`

**What is the expected behavior?**

Yarn doesn't install `webapp-webpack-plugin/node_modules/favicons` and uses declared and installed `favicons@5.1.1`, NPM works this way.

**Please mention your node.js, yarn and operating system version.**

yarn version v1.13.0-0, Windows 10

**Comment by whitecolor** [comment]
@bestander sorry to bother, but isn't it a basic case, or don't get something?

**Comment by bestander** [comment]
Yeah, this is a known problem.
Yarn sometimes does not optimize the dependency tree and it goes depth first from top to bottom generating it.
This is a great opportunity to improve the package resolution algorithm, this is not as basic problem actually when there are many dependencies it's hard to find the minimal possible set of dependencies that satisfy all requirements.

**Comment by whitecolor** [comment]
@bestander 

> Yarn sometimes does not optimize the dependency tree

Still looks like an inconsistency in terms of determination which yarn claims to achieve in the first place.

**Comment by bestander** [comment]
With yarn.lock the resolution is consistent.
The variability comes when Yarn resolves dependencies for the first time, package manager can't give guarantees here anyway because packages get updated over time and the shape of the tree would change anyway.

**Comment by telamonian** [comment]
I've run into a lot of related problems recently (see jupyterlab/jupyterlab-data-explorer#82). In terms of improving the package resolution algorithm, would it be enough to just: 

- track which packages get duplicate resolution in the `yarn.lock`
- for each package with duplicates, carefully reexamine all collected version ranges for that package and make the most parsimonious re-resolution

Or is that a dumb idea?

**Comment by CXuesong** [comment]
I was intending to open a duplicate issue, when I find this issue. I've experienced a similar situation recently,
* I installed `@types/webpack@4.41.2` and `@types/webpack-dev-server@3.10.1` some while ago. The latter has dependency (`@types/webpack@*`) on the former.
    * Generated `yarn.lock` looks like this, it's expected.
```yaml
"@types/webpack@*", "@types/webpack@^4.41.2":
  version "4.41.2"
  resolved "https://registry.yarnpkg.com/@types/webpack/-/webpack-4.41.2.tgz#c6faf0111de27afdffe1158dac559e447c273516"
  integrity sha512-DNMQOfEvwzWRRyp6Wy9QVCgJ3gkelZsuBE2KUD318dg95s9DKGiT5CszmmV58hq8jk89I9NClre48AEy1MWAJA==
  dependencies:
    "@types/anymatch" "*"
    "@types/node" "*"
    "@types/tapable" "*"
    "@types/uglify-js" "*"
    "@types/webpack-sources" "*"
    source-map "^0.6.0"
```
* Then after some time, I installed `@types/webpack@4.41.3`. Now `yarn.lock` looks like this
```yaml
"@types/webpack@*":
  version "4.41.2"
  resolved "https://registry.yarnpkg.com/@types/webpack/-/webpack-4.41.2.tgz#c6faf0111de27afdffe1158dac559e447c273516"
  integrity sha512-DNMQOfEvwzWRRyp6Wy9QVCgJ3gkelZsuBE2KUD318dg95s9DKGiT5CszmmV58hq8jk89I9NClre48AEy1MWAJA==
  dependencies:
    "@types/anymatch" "*"
    "@types/node" "*"
    "@types/tapable" "*"
    "@types/uglify-js" "*"
    "@types/webpack-sources" "*"
    source-map "^0.6.0"

"@types/webpack@^4.41.3":
  version "4.41.8"
  resolved "https://registry.yarnpkg.com/@types/webpack/-/webpack-4.41.8.tgz#d2244f5f612ee30230a5c8c4ae678bce90d27277"
  integrity sha512-mh4litLHTlDG84TGCFv1pZldndI34vkrW9Mks++Zx4KET7DRMoCXUvLbTISiuF4++fMgNnhV9cc1nCXJQyBYbQ==
  dependencies:
    "@types/anymatch" "*"
    "@types/node" "*"
    "@types/tapable" "*"
    "@types/uglify-js" "*"
    "@types/webpack-sources" "*"
    source-map "^0.6.0"
```
And I can see the following foler created in my project, i.e. nested `node_modules`:
```
node_modules\@types\webpack-dev-server\node_modules\@types\webpack\index.d.ts
```
basically causing issue like DefinitelyTyped/DefinitelyTyped#43232.

I think this can cause trouble when upgrading the dependencies. Hope this issue can be resolved...
